disperser/apiserver/server.go

```diff
@@ -76,42 +78,119 @@ func NewDispersalServer(
 	}
 }
 
+func (s *DispersalServer) DisperseBlobAuthenticated(stream pb.Disperser_DisperseBlobAuthenticatedServer) error {
+
+	// Process disperse_request
+	in, err := stream.Recv()
+	if err != nil {
+		return fmt.Errorf("error receiving next message: %v", err)
+	}
+
+	request, ok := in.Payload.(*pb.AuthenticatedRequest_DisperseRequest)
+	if !ok {
+		return errors.New("expected DisperseBlobRequest")
+	}
+
+	// Send back challenge to client
+	challenge := rand.Uint32()
+	err = stream.Send(&pb.AuthenticatedReply{Payload: &pb.AuthenticatedReply_Challenge{
+		Challenge: &pb.Challenge{
+			ChallengeParameter: challenge,
+		},
+	}})
+	if err != nil {
+		return err
+	}
+
+	// Recieve challenge_reply
+	in, err = stream.Recv()
+	if err != nil {
+		return fmt.Errorf("error receiving next message: %v", err)
+	}
+
+	challengeReply, ok := in.Payload.(*pb.AuthenticatedRequest_ChallengeReply)
+	if !ok {
+		return errors.New("expected ChallengeReply")
+	}
+
+	blob := getBlobFromRequest(request.DisperseRequest)
+
+	// TODO(mooselumph): Add auth data to blob object
+
+	receivedChallenge, err := strconv.ParseUint(string(challengeReply.ChallengeReply.AuthenticationData), 10, 32)
+	if err != nil {
+		return err
+	}
+	if uint32(receivedChallenge) != challenge+1 {
+		return fmt.Errorf("invalid challenge")
+	}
+
+	_ = challengeReply
+	// blob.RequestHeader.Nonce = challenge
+	// blob.RequestHeader.AuthenticationData = challengeReply.ChallengeReply.AuthenticationData
+
+	// Disperse the blob
+	reply, err := s.disperseBlob(stream.Context(), blob, true, challenge)
+	if err != nil {
+		return err
+	}
+
+	// Send back disperse_reply
+	err = stream.Send(&pb.AuthenticatedReply{Payload: &pb.AuthenticatedReply_DisperseReply{
+		DisperseReply: reply,
+	}})
+	if err != nil {
+		return err
+	}
+
+	return nil
+
+}
+
 func (s *DispersalServer) DisperseBlob(ctx context.Context, req *pb.DisperseBlobRequest) (*pb.DisperseBlobReply, error) {
+
+	blob := getBlobFromRequest(req)
+
+	return s.disperseBlob(ctx, blob, false, 0)
+
+}
+
+func (s *DispersalServer) disperseBlob(ctx context.Context, blob *core.Blob, authenticated bool, challengeParam uint32) (*pb.DisperseBlobReply, error) {
```

Comment:
How are `authenticated` and `challengeParam` used?

Reply:
Removed these for now.